### PR TITLE
MM-11069: Adds ability to view archived channels.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary
Adds ability to view archived channels.

#### Ticket Link
[MM-11069](https://mattermost.atlassian.net/browse/MM-11069)

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (https://github.com/mattermost/mattermost-server/pull/9039)
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates
- [x] Touches critical sections of the codebase (auth, posting, etc.)

### CODE REVIEW

**Comment by mkraft** [comment]
@esethna Sounds good.

@jespino Note that the API endpoint to leave a channel must remain operable for an archived channel.

**Comment by mkraft** [comment]
Front-end now allows one to 'leave channel' and it should work once the API is updated to allow it by @jespino.

**Comment by mkraft** [comment]
@jespino I have added a bunch more snaps.

File: sass/layout/_headers.scss [hunk 1]
```diff
@@ -703,3 +703,7 @@
     }
 
 }
+
+.channel-header-archived-icon {
+    margin-right: 5px;
+}
```
Comment (cpanato) [inline sass/layout/_headers.scss]: missing end blank line :)

File: tests/components/permalink_view.test.jsx [hunk 2]
```diff
@@ -40,4 +40,15 @@ describe('components/PermalinkView', () => {
...
+        const props = {...baseProps, channelIsArchived: true};
+
+        const wrapper = shallow(
+            <PermalinkView {...props}/>
+        );
+
+        wrapper.setState({valid: true});
+        expect(wrapper).toMatchSnapshot();
+    });
 });
```
Comment (cpanato) [inline tests/components/permalink_view.test.jsx]: here as well :)

**Comment by esethna** [comment]
Thanks Martin, a few tweaks:

1. Change this text (only when the archive search config is on), to

```
This will archive the channel from the team. Channel contents will still be accessible by channel members.

Are you sure you wish to archive the nostrum channel?
```

image

2. CSS looks a little funcky here when the channel width is small. @asaadmahmood? 

image

3. I remember we discussed going to the last viewed channel when closing an archived channel. Right now it goes to town square which is a sub-par user experience.

**Comment by asaadmahmood** [comment]
Channel name not displayed for archived channel in a certain case.
Repro steps:
1. Type a message in a channel.
2. Search for that message.
3. Have that message visible in search.
4. Archive that channel.
5. Observe that the archived tags pops up but the channel name dissapears.
screen shot 2018-08-01 at 12 50 09 am

**Comment by asaadmahmood** [comment]
@mkraft Is there a reason we have used svgs instead of just using these icons as fonts via font awesome?

**Comment by asaadmahmood** [comment]
Refreshing while on the archived channel brings you to an error page. (Page not found).

**Comment by asaadmahmood** [comment]
Searching for a message in an archived channel does not work if I archived it before searching.
I can search a message, it appears in my RHS search, I can then archive the channel, it remains in the RHS, and then I can jump to it, but locally. I can't simply search for a message in an archived channel after archiving it.

**Comment by asaadmahmood** [comment]
@esethna Pushed my UI changes, along with some other things that weren't mentioned.

**Comment by mkraft** [comment]
@asaadmahmood Regarding your (good) catch as described here: https://github.com/mattermost/mattermost-webapp/pull/1415#issuecomment-409346105, I have opened a ticket https://mattermost.atlassian.net/browse/MM-11546.

**Comment by mkraft** [comment]
@esethna I have changed the archive warning modal.

**Comment by mkraft** [comment]
@asaadmahmood Reply to https://github.com/mattermost/mattermost-webapp/pull/1415#issuecomment-409346668: No particular reason other than I didn't know about another way of doing it.

@jespino Were you planning to change the API to address https://github.com/mattermost/mattermost-webapp/pull/1415#issuecomment-409349892 or were you expecting a front-end change?

reply to https://github.com/mattermost/mattermost-webapp/pull/1415#issuecomment-409351635: I'm not able to reproduce, it works fine for me.

**Comment by jespino** [comment]
@mkraft I would say is a frontend issue. the backend already expose the deleted channel by name, but you have to ask it with the "include_deleted" query parameter.

**Comment by mkraft** [comment]
@jespino Yes it was being caused by the join-channel flow which I have addressed with https://github.com/mattermost/mattermost-webapp/pull/1415/commits/630abe40f8c23e9a862a7637b8965455aa852391.

**Comment by esethna** [comment]
One ticket in progress: https://mattermost.atlassian.net/browse/MM-11529 But not going to block merge

**Comment by esethna** [comment]
Reassigning as Katie is working on the docs for this feature

**Pull request merged**